pulsar/consumer_partition.go

```diff
@@ -571,6 +627,114 @@ func (pc *partitionConsumer) MessageReceived(response *pb.CommandMessage, header
 	return nil
 }
 
+func createEncryptionContext(msgMeta *pb.MessageMetadata) EncryptionContext {
+	encCtx := EncryptionContext{
+		Algorithm:        msgMeta.GetEncryptionAlgo(),
+		Param:            msgMeta.GetEncryptionParam(),
+		UncompressedSize: int(msgMeta.GetUncompressedSize()),
+		BatchSize:        int(msgMeta.GetNumMessagesInBatch()),
+	}
+
+	if msgMeta.Compression != nil {
+		encCtx.CompressionType = CompressionType(*msgMeta.Compression)
+	}
+
+	kMap := map[string]EncryptionKey{}
+	for _, k := range msgMeta.GetEncryptionKeys() {
+		metaMap := map[string]string{}
+		for _, m := range k.GetMetadata() {
+			metaMap[*m.Key] = *m.Value
+		}
+
+		kMap[*k.Key] = EncryptionKey{
+			KeyValue: k.GetValue(),
+			Metadata: metaMap,
+		}
+	}
+
+	encCtx.Keys = kMap
+	return encCtx
+}
+
+func (pc *partitionConsumer) decryptPayLoadIfNeeded(msgID *pb.MessageIdData,
+	msgMeta *pb.MessageMetadata,
+	payload []byte) ([]byte, error) {
+
+	// No encryption keys found in message metadata, do not decrypt payload
+	if len(msgMeta.GetEncryptionKeys()) == 0 {
+		return payload, nil
+	}
+
+	// If KeyReader interface is not implemented, throw an error based on the configuration
+	if pc.options.KeyReader == nil {
+		switch pc.options.consumerCryptoFailureAcrion {
+		case crypto.Consume:
+			pc.log.Warnf("[%v][%v][%v] KeyReader interface is not implemented. Consuming encrypted message.",
+				pc.topic,
+				pc.options.subscription,
+				pc.options.consumerName,
+			)
+			return payload, errKeyReaderNotImplemted
+		case crypto.Discard:
+			pc.log.Warnf(`[%v][%v][%v] Skipping decryption since KeyReader 
+			interface is not implemented and config is set to discard`,
+				pc.topic,
+				pc.options.subscription,
+				pc.options.consumerName,
+			)
+			pc.discardCorruptedMessage(msgID, pb.CommandAck_DecryptionError)
+			return nil, errKeyReaderNotImplemted
+		case crypto.FailConsume:
+			pc.log.Errorf(
+				`[%v][%v][%v][%v] Message delivery failed since KeyReader interface 
+				is not implemented to consume encrypted message`,
+				pc.topic,
+				pc.options.subscription,
+				pc.options.consumerName,
+				msgID,
+			)
+			return nil, errKeyReaderNotImplemted
+		}
+	}
+
+	decryptedPayload, err := pc.options.messageCrypto.Decrypt(msgMeta, payload, pc.options.KeyReader)
+	if err != nil {
+		errMsg := fmt.Errorf(errDecryptingPayload, err.Error())
+		switch pc.options.consumerCryptoFailureAcrion {
+		case crypto.FailConsume:
+			pc.log.Warnf(
```

Comment:
This info about the consumer will already be logged
```
pc.log = client.log.SubLogger(log.Fields{
		"name":         pc.name,
		"topic":        options.topic,
		"subscription": options.subscription,
		"consumerID":   pc.consumerID,
	})
```

Reply:
Ok. I'll remove the redundant info